falco-driver-loader - break apart logic and improvements

leogr opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines in the [CONTRIBUTING.md](CONTRIBUTING.md) file and learn how to compile Falco from source [here](https://falco.org/docs/source).
2. Please label this pull request according to what type of issue you are addressing.
3. . Please add a release note!
4. If the PR is unfinished while opening it specify a wip in the title before the actual title, for example, "wip: my awesome feature"
-->

**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

/kind feature

> If contributing rules or changes to rules, please make sure to also uncomment one of the following line:

> /kind rule-update

> /kind rule-create

<!--
Please remove the leading whitespace before the `/kind <>` you uncommented.
-->

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area build

> /area engine

> /area examples

> /area rules

> /area integrations

> /area tests

> /area proposals

<!--
Please remove the leading whitespace before the `/area <>` you uncommented.
-->

**What this PR does / why we need it**:
See #1193

**Which issue(s) this PR fixes**:

<!--
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
If PR is `kind/failing-tests` or `kind/flaky-test`, please post the related issues/tests in a comment and do not use `Fixes`.
-->

Fixes #1193
Fixes #1125 (again)

**Special notes for your reviewer**:

```shell
Usage:
  falco-driver-loader [driver] [options]

Available drivers:
  module        kernel module (default)
  bpf           eBPF probe

Options:
  --help        show brief help
  --compile     try to compile the driver locally
  --download    try to download a prebuilt driver
```

With no options and arguments, it preserves the previous behavior (we should test it deeply).

Notable improvements in messages:
 - All info messages start with `* Message...`
 - Success messages with `* Success: message...`
 - Messages without `*` represent the output of another program or an error
 - Verbosity increased 

/cc @kris-nova 
/cc @leodido 
/cc @fntlnz 

### falco-driver-loader (pre refactoring behavior)

Preflight check:
- if not root, then exit 1
- if not `curl`, then exit 1

#### module

Preflight check:
- if not `lsmod`, then exit 1
- if not `modprobe`, then exit 1
- if not `rmmod`, then exit 1


1. Unload module, if present
2. If module still loaded, then exit 0
3. If UEK host, then
    - do nothing
    *else*
    - if `dkms install` and `insmod` then exit 0
4. if `modprobe` module, then exit 0
5. if unsupported OS, then exit 1 (see `get_target_id`)
6. if `~/.falco/falco_<target_id>_<kernel_release>_<kernel_version>.ko`, then:
    - if `insmod` then exit 0, else exit 1
7. if `curl` module, then
    - if `insmod` then exit 0, else exit 1
    *else*
    - exit 1
8. exit 0

#### bpf 

Preflight check:
- Mount debugfs (continue on error)
- if kernel config not found, exit 1 (see `get_kernel_config`)
- if unsupported OS, then exit 1 (see `get_target_id`)

1. if NOT `~/.falco/falco_<target_id>_<kernel_release>_<kernel_version>.o`, then go to step 6.
    - if COS, MINKUBE, or `BPF_USE_LOCAL_KERNEL_SOURCES` then download kernel source
        - if download failed, then exit 1
        - `customize_kernel_build`
    - make bpf probe to `~/.falco/falco_<target_id>_<kernel_release>_<kernel_version>.o`
2. if NOT `~/.falco/falco_<target_id>_<kernel_release>_<kernel_version>.o`, 
    - `curl` bpf probe (continue on error)
3. if `~/.falco/falco_<target_id>_<kernel_release>_<kernel_version>.o`, then
    - if symlink to `~/.falco/falco-bpf.o` then exit 0, else exit 1
    *else*
    - exit 1
4. exit 0

**N.B.**: **bpf** exists with 0 even though `curl` failed (see step 2.)

**Does this PR introduce a user-facing change?**:

<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below.
If the PR requires additional action from users switching to the new release, prepend the string "action required:".
For example, `action required: change the API interface of the rule engine`.
-->

```release-note
update(scripts): improve `falco-driver-loader` output messages
new(scripts): options and command-line usage for `falco-driver-loader`
```

poiana commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **fntlnz**
You can assign the PR to them by writing `/assign @fntlnz` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=falcosecurity%2Ffalco).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/falcosecurity/falco/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["fntlnz"]} -->

leodido commented on scripts/falco-driver-loader:
```diff
@@ -420,18 +480,100 @@ fi
 	exit 1
 fi
 
 if ! hash curl > /dev/null 2>&1; then
```
Can you please add a check for `dkms` when user wants to compile?

leogr replied:
It is already present [here](https://github.com/falcosecurity/falco/pull/1200/files#diff-b1a242527066116a4e7485802c03bbb3R154) (see the `else` branch).
The check will not terminate the execution as like as the previous script version.

leogr replied:
Furthermore, we need to continue the execution so the download strategy can happen.
When `dkms` does not exist, users will see `* Skipping dkms install (dkms not found)`

leogr commented on scripts/falco-driver-loader:
```diff
@@ -140,6 +139,84 @@ get_target_id() {
+	fi
+
+	if [ -n "$SKIP_LOAD" ]; then
+		if modprobe "${DRIVER_NAME}" > /dev/null 2>&1; then
```
It should be a `--dry-run`

leodido commented on scripts/falco-driver-loader:
```diff
@@ -420,18 +480,100 @@ fi
+	    has_opts="true"
+      shift
+      ;;
+	--skip-load)
```
Indentation

leodido commented on scripts/falco-driver-loader:
```diff
@@ -420,18 +480,100 @@ fi
+      exit 0
+      ;;
+    --compile)
+      	ENABLE_COMPILE="yes"
```
Indentation

leodido commented on scripts/falco-driver-loader:
```diff
@@ -177,226 +241,222 @@ load_kernel_module() {
+		if [ -f "${HOME}/.falco/${BPF_PROBE_FILENAME}" ]; then
+			echo "* Skipping download, eBPF probe is already present in ${HOME}/.falco/${BPF_PROBE_FILENAME}"
+		else
+			load_bpf_probe_download
```
Check the operation was successful so to give hints to the user

leogr replied:
Fixed here :point_right: https://github.com/falcosecurity/falco/pull/1200/files#diff-b1a242527066116a4e7485802c03bbb3R378

leogr commented:
> The only thing I'm not really sure about is the usage of `check_skip_load` to circuit-break only the load phase of the drivers (not their installation).
> 
> So, implemented in this way, what's the use case of `--skip-load`? Maybe I'm missing something..

My thought when I saw the proposal in the issue was... we might need this for testing purposes only. Ofc, Falco will not work if `--skip-load` is used.

Should we remove it?

leogr commented:
Hey @fntlnz 
I'm working on tests, so we can add them.

That being said, really important design question: after talking with @leodido I believe we should remove the `--skip-load` option. WDYT?
If you agree on to remove it too, I will remove it ASAP.

leodido commented:
I think `--skip-load` flag should be removed: it adds more complexity than it removes. 

The `--download` and `--compile` flags are pretty explanatory and they cover the use cases

My 2 cents

leogr commented:
Update:
- `--skip-load` has been removed
- do not continue when epbf download fails

leogr commented:
**IMPORTANT**

I have noticed that:
```
	echo "* Trying to load a system ${DRIVER_NAME} driver, if present"
	if modprobe "${DRIVER_NAME}" > /dev/null 2>&1; then
		echo "* Success: ${DRIVER_NAME} module found and loaded with modprobe"
		exit 0
	fi
```
might load a kernel module that does not fit the installed Falco version. Currently, it happens before trying a local prebuilt and the download strategy, but that should be the last resort.

leogr commented:
Rebased on top of master to include the fix introduced by https://github.com/falcosecurity/falco/pull/1212

I removed WIP, because it's ok for me, though it's not possible to test every single case.

@fntlnz @leodido So, question: do we want this PR to be merged for the 0.23.0?

leogr commented:
Rebased on top of master to include lastest fixes.

leogr commented:
/milestone 0.23.0

poiana commented:
LGTM label has been added.  <details>Git tree hash: 947bf6a4f4a5ca071a6c0ec2bb56c12d1edb4971</details>

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/falco/pull/1200#pullrequestreview-413574538" title="Approved">fntlnz</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=falcosecurity%2Ffalco).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/falco/blob/master/OWNERS)~~ [fntlnz]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/falco/pull/1200#pullrequestreview-413574538" title="Approved">fntlnz</a>*, *<a href="https://github.com/falcosecurity/falco/pull/1200#pullrequestreview-413580222" title="Approved">leodido</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=falcosecurity%2Ffalco).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/falco/blob/master/OWNERS)~~ [fntlnz,leodido]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->
